### PR TITLE
feat: Copilot

### PR DESCRIPTION
**Pull request by petarvujovic98**
Add NEAR copilot into marketplace
Add copilot page
Add chat API
Add additional hooks to support chat functionality
Refactor copilot code

### CODE REVIEW

**Comment by petarvujovic98** [comment]
@kenobijon I need access to the Mendable API key in order to deploy this preview and to be able to test it out, other than that it should be fully integrated apart from any visual changes that would arise due to chat testing.

File: app/src/app/copilot/chat/bubble.tsx [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+}) {
+  const { sources, messageWithOutSources, messageId } = getSources(message);
+
+  return (
+    <div
+      key={message.id}
+      className={`flex flex-col items-center border-none ${
+        message.role === "user"
+          ? "bg-human-background text-human-foreground border-b border-gray-900/50"
+          : "bg-assistant-background text-assistant-foreground border-b border-gray-900/50"
```
Comment (kenobijon) [inline app/src/app/copilot/chat/bubble.tsx]: Non-Blocking: Add in css for assistant and human text blocks

File: app/src/app/copilot/chat/rate-message.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
+"use client";
+
+import { useCallback, useState } from "react";
+import { useRateMessage } from "~/hooks/copilot";
+import { type MessageRating } from "~/lib/validation/copilot";
+
+export function RateMessage({ id }: { id: number }) {
```
Comment (kenobijon) [inline app/src/app/copilot/chat/rate-message.tsx:7]: 404 (Not Found) Error occuring when rateMessage is called

File: app/src/app/copilot/markdown/react-markdown.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                  {children}
+                </a>
+                {!lastChild.trim().endsWith(".") && ". "}
+              </>
+            );
+          }
+
+          return <>{children}</>;
+        },
+      }}
```
Comment (kenobijon) [inline app/src/app/copilot/markdown/react-markdown.tsx]: This doesn't seem to be getting rendered properly. This was to solve a corner case for markdown links in order to add proper punctuation. The link doesn't seem to be getting generated along with the punctuation

File: app/src/app/copilot/page.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                  loading={isLoading}
+                />
+              )}
+            </ScrollArea>
+          </ChatContent>
+        </>
+      )}
+
+      <ChatFooter className="sticky bottom-0 w-full space-x-2 pt-2">
+        <div className="content_width flex items-center bg-transparent">
```
Comment (kenobijon) [inline app/src/app/copilot/page.tsx]: Non-Blocking: Future fix to make the text box larger and include expanding the text box on overflow so the user can view the entire prompt

**Review by kenobijon — COMMENTED** [review]
Approval upon fix of rateMessage. 

Can incorporate some of the css for the chatbot in the future.

File: app/src/app/copilot/chat/rate-message.tsx (hunk 2 above)
Comment (petarvujovic98) [reply app/src/app/copilot/chat/rate-message.tsx:7]: Solved by [eb77b9b](https://github.com/near-horizon/horizon/pull/72/commits/eb77b9ba0860864732d9520c40fecc7e86162087)

I was sending the call to an incorrect route. It now is corrected

File: app/src/app/copilot/chat/bubble.tsx (hunk 1 above)
Comment (petarvujovic98) [reply app/src/app/copilot/chat/bubble.tsx]: Solved by [e246a83](https://github.com/near-horizon/horizon/pull/72/commits/e246a83b27c259d97f8572b24ad6ddaae6c0cf21)

This should be revisited to improve styling as it looks off compared to the rest of the app.

File: app/src/app/copilot/markdown/react-markdown.tsx (hunk 3 above)
Comment (petarvujovic98) [reply app/src/app/copilot/markdown/react-markdown.tsx]: What is the exact edge case? Could you describe what the link would look like so I could properly extract it and generate the replacement?

File: app/src/app/copilot/page.tsx (hunk 4 above)
Comment (petarvujovic98) [reply app/src/app/copilot/page.tsx]: Does this mean you want to make the input expand vertically like a text area to show the whole prompt? If so, what do we want the max number of lines to be?

File: app/src/app/copilot/markdown/react-markdown.tsx (hunk 3 above)
Comment (kenobijon) [reply app/src/app/copilot/markdown/react-markdown.tsx]: Screenshot 2023-12-18 at 1 41 15 PM

File: app/src/app/copilot/page.tsx (hunk 4 above)
Comment (kenobijon) [reply app/src/app/copilot/page.tsx]: possibly 3-4

**Comment by kenobijon** [comment]
Looks good to me following small UI improvements discussed on prompt cards

**Comment by kenobijon** [comment]
Looks good to me!

**Review by kenobijon — APPROVED** [review]

**Pull request merged**